### PR TITLE
[FIX] Protein terminal mods in AASequence

### LINKED ISSUE #4830: Variable Modification "Met-loss (Protein N-term M)" yields error in CometAdapter
This error is thrown:
```
Error: Unexpected internal error (the value 'Met-loss (Protein N-term M)' was used but is not valid; Retrieving the modification failed. It is not available for the residue '' and term specificity 'N-term'. )
Error occurred in line 265 of file C:\Development\OpenMS\src\openms\source\CHEMISTRY\ModificationsDB.cpp (in function: const class OpenMS::ResidueModification *__cdecl OpenMS::ModificationsDB::getModification(const class OpenMS::String &,const class OpenMS::String &,enum OpenMS::ResidueModification::TermSpecificity) const) !
```

This modification is supported by Comet as the 'clip_nterm_methionine' parameter in its param file and can also be passed to CometAdapter as a flag.
Maybe CometAdapter should check for this modification in the 'variable_modifications' parameter and set 'clip_nterm_methionine' accordingly instead of writting the modification.

### PR DESCRIPTION
**Pull request by jpfeuffer**
# Description
Fixes #4830 
Fixes bug in AASequence when trying to set terminal modifications.
Now correctly parses residue-specific terminal mods.
And if not found for general terminal mods, it automatically looks for protein terminal mods.
The best way in the future would be to actually check if this peptide is protein terminal and only look up the corresponding modifications but this logic needs to be outside of AASequence and implemented everywhere where we have access to this information.

# Checklist:
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] New and existing unit tests pass locally with my changes
- [x] Updated or added python bindings for changed or new classes

### CODE REVIEW

**Comment by jpfeuffer** [comment]
What the hell?

**Comment by jpfeuffer** [comment]
I did not touch anything related to XTandem. And the converted mzMLs are fine as seen on non-Windows.

**Comment by jpfeuffer** [comment]
@cbielow Can you reproduce this?

**Comment by cbielow** [comment]
Yes, can reproduce Comet's error.
When doing `FileConverter.exe -in C:/dev/openms_test/share/OpenMS/examples/FRACTIONS/BSA1_F1.mzML -out C:/dev/openms_test/share/OpenMS/examples/FRACTIONS/BSA1_F1.mzML`, then Comet does not crash anymore and finished successfully.. although the results are slightly different (suggesting that on Linux it also does something weird, but does not crash).

XTandem I cannot reproduce (X! TANDEM Vengeance (2015.12.15.2)).


btw, since it came up in the tests (printing `(exit code: ù).`): could you change
`(exit code: " + qp_->exitCode() + "). ` to `(exit code: " + int(qp_->exitCode()) + "). ` in
https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/SYSTEM/ExternalProcess.cpp#L117
while you're at it?
Thanks!

**Comment by jpfeuffer** [comment]
Weird. That Fileconversion command was exactly what I did for the files. This is why they changed.

**Comment by jpfeuffer** [comment]
I just converted again. Results in the same mzML file (except for another dataprocessing entry) and updated offsets.
Same results in all tests.
I think Windows Comet just can't read non-Windows generated things.

**Comment by jpfeuffer** [comment]
rebuild jenkins

**Comment by cbielow** [comment]
solves https://github.com/OpenMS/OpenMS/issues/4654 (Comet crashes due to conversion of line endings from LF to CRLF by Git, which invalidates the mzML indeX)

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 1]
```diff
@@ -1429,23 +1429,53 @@ namespace OpenMS
...
-    if (modification == "")
+    if (modification.empty())
     {
       n_term_mod_ = nullptr;
       return;
     }
 
-    n_term_mod_ = ModificationsDB::getInstance()->getModification(modification, "", ResidueModification::N_TERM);
+    String residue = "";
+    char last_char_no_parentheses = modification[modification.length()-2];
```
Comment (hroest) [inline src/openms/source/CHEMISTRY/AASequence.cpp:1439]: what happens if I enter a string of size 1? 

```
setNTerminalModification("T")
```

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 2]
```diff
@@ -1429,23 +1429,53 @@ namespace OpenMS
...
-    if (modification == "")
+    if (modification.empty())
     {
       c_term_mod_ = nullptr;
       return;
     }
-    c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, "", ResidueModification::C_TERM);
+
+    String residue = "";
+    char last_char_no_parentheses = modification[modification.length()-2];
```
Comment (hroest) [inline src/openms/source/CHEMISTRY/AASequence.cpp:1465]: same here

File: src/openms/source/CHEMISTRY/AASequence.cpp (hunk 1 above)
Comment (hroest) [reply src/openms/source/CHEMISTRY/AASequence.cpp:1439]: @jpfeuffer can you please add that to the AASequence_test.cpp

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 3]
```diff
@@ -1445,9 +1444,19 @@ namespace OpenMS
...
+    // For strings in our most common UniMod format
     if (!modification.hasSubstring("Protein N-term"))
     {
-      n_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::N_TERM);
+      try
+      {
+        n_term_mod_ = ModificationsDB::getInstance()
+            ->getModification(modification, residue, ResidueModification::N_TERM);
+      }
+      catch (...)
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/AASequence.cpp:1455]: please comment try catch logic

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 4]
```diff
@@ -1476,7 +1485,14 @@ namespace OpenMS
 
     if (!modification.hasSubstring("Protein C-term"))
     {
-      c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::C_TERM);
+      try
```
Comment (timosachsenberg) [inline src/openms/source/CHEMISTRY/AASequence.cpp:1488]: same ehere

**Comment by timosachsenberg** [comment]
do the pxds need an update?

**Comment by jpfeuffer** [comment]
Unless pxds need special annotation when you catch exceptions, no. No interface changed.. yet.
It should heavily though, to make this pepXML parsing and general mod. handling from a nightmare to usable.

Failures unrelated.. I really need to check why their frequency increased after the update.

**Comment by timosachsenberg** [comment]
I thought there were some references introduced that might be supported

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 5]
```diff
@@ -1429,23 +1428,84 @@ namespace OpenMS
...
+    if (!modification.hasSubstring("Protein C-term"))
+    {
+      // since this method is called setNTerminalModification without further specification
+      // we have to look for both general terminus and Protein terminus.
+      // For backwards compatibility we look for the general terminus first
+      // We have to use try-catch since getModification unfortunately throws Exceptions.
+      try
+      {
+        c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::C_TERM);
+      }
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/AASequence.cpp:1499]: return from function here?

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 6]
```diff
@@ -1429,23 +1428,84 @@ namespace OpenMS
...
+      catch (...)
+      {
+        c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::PROTEIN_C_TERM);
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/AASequence.cpp:1502]: catch, but don't do anything.

File: src/openms/source/CHEMISTRY/AASequence.cpp [hunk 7]
```diff
@@ -1429,23 +1428,84 @@ namespace OpenMS
...
+      }
+    }
+    else
+    {
+      c_term_mod_ = ModificationsDB::getInstance()->getModification(modification, residue, ResidueModification::PROTEIN_C_TERM);
```
Comment (cbielow) [inline src/openms/source/CHEMISTRY/AASequence.cpp:1507]: and then always do this.

Would make the code easier and less repetitive?

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 8]
```diff
@@ -1319,31 +1347,71 @@ namespace OpenMS
...
+      }
+
+      // BIG NOTE: According to the pepXML schema specification, protein terminus is either "c" or "n" if set.
+      // BUT: Many tools will put "Y" or "N" there in conjunction with the terminus attribute.
+      // Unfortunately there is an overlap for the letter "n". We will try to handle both based on the case:
+      String entry_lower = protein_terminus_entry;
+      entry_lower = entry_lower.toLower();
+      if (entry_lower == "y")
+      {
+        aa_mod.protein_terminus = true;
```
Comment (cbielow) [inline src/openms/source/FORMAT/PepXMLFile.cpp:1367]: sorry, but all this code does not belong here!

collect all the information you need and refactor it to `aa_mod`'s implemention.
Ideally, make it a constructor and forbid changing properties individually (to maintain invariants).

This is a typical case of 'just add the code inline' which makes it
 - not testable
 - creates huge unmaintainable classes
 - is no fun to refactor at all

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 9]
```diff
@@ -1365,6 +1433,7 @@ namespace OpenMS
       if (desc.empty())
       {
         vector<String> mods;
+        // if terminus was not specified
```
Comment (cbielow) [inline src/openms/source/FORMAT/PepXMLFile.cpp:1436]: creating a string from `aa_mod` should be a member function of aa_mod!

File: src/openms/source/FORMAT/PepXMLFile.cpp (hunk 8 above)
Comment (jpfeuffer) [reply src/openms/source/FORMAT/PepXMLFile.cpp:1367]: aa_mod should not be here at all.
It is:
a) abused also for terminal mods
b) should be our own Residuemodification object, such that we do not lookup ever changing "mod-descriptions" dozens of times

**Comment by jpfeuffer** [comment]
I have to give up now. Not sure when I get to it again.

File: src/openms/source/FORMAT/PepXMLFile.cpp [hunk 10]
```diff
@@ -1863,7 +1863,8 @@ namespace OpenMS
 
       for (const auto& mod : fixed_modifications_)
       {
-        if (mod.getTerminus() == "n")
+        if (mod.getRegisteredMod()->getTermSpecificityName() == "N-term" ||
```
Comment (cbielow) [inline src/openms/source/FORMAT/PepXMLFile.cpp:1866]: `.hasSuffix("N-term")`  to make it shorter?

**Pull request merged**